File: src/main/java/org/hust/audioSearch/shazam/Fingerprint.java
package org.hust.audioSearch.shazam;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

public class Fingerprint {
    public static Map<Integer, ArrayList<Couple>> generateFingerprints(Peak[]peaks, int songId){
        Map<Integer,ArrayList<Couple>> fingerprints = new HashMap<>();
        int targetZoneSize = 5;
        for (int anchorIndex = 0; anchorIndex < peaks.length-2-targetZoneSize; anchorIndex++) {
            for (int pointIndex = anchorIndex+3; pointIndex < anchorIndex+ 3+targetZoneSize; pointIndex++) {
                int address = 0;
                Peak anchor = peaks[anchorIndex];
                Peak point = peaks[pointIndex];
                address = address|(anchor.frequency&0x1ff);
                address = address<<9|(point.frequency&0x1ff);
                address = address<<14|((int)(point.time - anchor.time)&0x3fff);
                fingerprints.computeIfAbsent(address, key -> new ArrayList<>())
                        .add(new Couple( (int)(anchor.time), songId));
//                System.out.println(String.format("%s, %s, %s", anchor.frequency, point.frequency, point.time - anchor.time));
            }
        }
        return fingerprints;
    }
}

File: FileManagementService/audioRecognition/service/Fingerprint.js
const { Couple } = require('../model/Couple')

function generateFingerprints(peaks, songId) {
    const fingerprints = new Map();
    const targetZoneSize = 5;

    for (let anchorIndex = 0; anchorIndex < peaks.length - 2 - targetZoneSize; anchorIndex++) {
        for (let pointIndex = anchorIndex + 2; pointIndex < anchorIndex + 2 + targetZoneSize; pointIndex++) {
            let address = 0;
            const anchor = peaks[anchorIndex];
            const point = peaks[pointIndex];

            address |= anchor.frequency & 0x1ff; // Mask the frequency to the lower 9 bits
            address <<= 9;
            address |= point.frequency & 0x1ff; // Mask the frequency to the lower 9 bits
            address <<= 14;
            address |= (point.time - anchor.time) & 0x3fff; // Mask the time difference to the lower 14 bits

            // Check if the address already exists in the map, if not, initialize it
            if (!fingerprints.has(address)) {
                fingerprints.set(address, []);
            }

            // Add the Couple object to the list corresponding to the address
            fingerprints.get(address).push(new Couple(anchor.time, songId));
        }
    }

    return fingerprints;
}

module.exports = {
    generateFingerprints
}